### PR TITLE
Highlighting and indentation

### PR DESCRIPTION
**Pull request by alezost**
Hello, this is another set of indentation & highlighting fixes.  But
please don't merge it now, I'd like to have a discussion on the
font-lock highlighting policy (see below).

1. `match-let` can have a "named" form as `let`, i.e. it can look like this:

   ```scheme
   (match-let loop (...)
     body)
   ```

   While currently it is indented like this:

   ```scheme
   (match-let loop (...)
              body)
   ```

   Sorry, it is my bad (introduced in commit 424553e).

2. The main thing: since `map` is highlighted (by `scheme-mode`), I
   think it would also be good to highlight `filter-map` and
   `append-map`.  But also there are `map!` and `append-map!`.
   Highlight these keywords as well?

   Also there are `fold` and `fold-right`.  And what about `unfold`,
   `unfold-right`, `reduce`, `reduce-right`, `filter`, `partition`,
   `remove` and other things?  I don't have an opinion on these.  On one
   hand `fold` has the same rights as `map`, but OTOH wouldn't it be too
   much if all such procedures were highlighted?

   Also currently `set!` is also font-locked (this highlighting was moved from
   chicken: commit 66f90b1).  I don't like it.

   To sum up: I vote to add highlighting for: `map!`, `append-map`,
   `append-map!` and `filter-map`.  And to remove highlighting for `set!`.
   What do you think?

### CODE REVIEW

**Comment by jaor** [comment]
hi alex, sorry for the late reply.

i'm more or less neutral about map!, append-map(!), but i think we
should definitely have highlighting for set! ... it's been in all scheme
standards i know, and i see no reason whatsover to remove it: what's the
rationale? am i missing something?

**Comment by alezost** [comment]
Hi!  Actually I don't understand why some keywords are highlighted and
some are not.  Yes, `set!` is the standard thing, but `list`, `cons`,
`car`, `cdr`, `pair?`, `append`, etc. are standard as well, but they are
not highlighted.

I think I don't like that `set!` is highlighted just because elisp's
`setq` is not highlighted.

Do you know if there is any rule what should be highlighted?

If not, there always be different opinions on what to highlight.  So I
think I better remove this controversial commit.

**Comment by jaor** [comment]
ah, i know see your rationale! yes, initially, i didn't want to
highlight standard functions, only syntax ("keywords"), because i
personally dislike too much fontlocking.  so what i did was to, mostly,
leave the problem to scheme-mode: whatever it highlights was fine by me.

if car, cdr &co. are *not* highlighted, i'm with you: it's more
consistent to remove set! too.  but then, perhaps we should also remove
highlighting for map instead of adding it for map! and friends.

but, on the other hand, i'm pretty sure people will start complaining if
we remove highlights from scheme-mode that they like, so perhaps we
should go in the other direction, leave set! there, and *add* car, cdr,
cons, list in the same way we're thinking map!, etc. should be added?

**Comment by alezost** [comment]
Oof, I think highlighting car, cdr and others will be too much.

I agree that removing highlighting for 'map' is not desired as people
got used to it (also I don't see how it can be done as it is defined in
scheme-mode).

All this highlighting stuff is too controversial.  Let's just leave it
all as it is now.  I've removed this highlighting commit, sorry for
bothering :-)

**Comment by jaor** [comment]
remaining patch cherry-picked and merged!
